[FLINK-10163] [sql-client] Support views in SQL Client

twalthr opened the pull request:
## What is the purpose of the change

Adds initial support for views in SQL Client. It adds the following statements:
- `CREATE VIEW`: Creates a virtual table from a SQL query. Syntax: 'CREATE VIEW <name> AS <query>'
- `SHOW VIEW`: Describes a previously created virtual table. Syntax: 'SHOW VIEW <name>'
- `DROP VIEW`: Deletes a previously created virtual table. Syntax: 'DROP VIEW <name>'

It also adds the section `views` to environment files.


## Brief change log

- Rewritten statement parser for statements with more than 1 operand
- View support added throughout the SQL Client
- Initial validation of environment files during startup
- Code clean up


## Verifying this change

- `SqlCommandParserTest`
- `ExecutionContextTest`
- `LocalExecutorITCase`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? docs

fhueske commented on flink-libraries/flink-sql-client/src/main/java/org/apache/flink/table/client/cli/CliClient.java:
```diff
@@ -390,51 +395,101 @@ private void callSelect(SqlCommandCall cmdCall) {
+
+		try {
+			// validate with a copy
+			final SessionContext contextCopy = context.copy();
```
This check fails if the VIEW query contains UDFs. 
It seems that these are not registered yet.

twalthr commented:
Thank you @fhueske for you feedback. I addressed your comments and improved the performance. I would discuss the listing of all views in a follow up issue.

fhueske commented on docs/dev/table/sqlClient.md:
```diff
@@ -459,6 +466,50 @@ Web interface: http://localhost:8081
+CREATE VIEW MyNewView AS SELECT MyField2 FROM MyTableSource
+{% endhighlight %}
+
+The `SHOW VIEW` statement allows for printing a previously created view again:
```
OK, I don't mind listing views are tables in `SHOW TABLES`.

However, if there is no way to distinguish a view from a table, we can also remove the `SHOW VIEW xxx` command. Why should somebody try to look up the definition of a view if he's not aware that it is a view. He can only know that if he created the view himself and in that case probably know the definition.

I also think the syntax of `SHOW VIEW xxx` is not well chosen because it is inconsistent with `SHOW TABLES` and `SHOW FUNCTIONS`. It gives *details about a single view* where as the other `SHOW` commands *list all elements of a type*.  I think the semantics are more similar to the `DESCRIBE` command. We could rename `SHOW VIEW` to ``DESCRIBE VIEW`.

twalthr replied:
FYI: In MySQL it is called [SHOW CREATE VIEW](https://dev.mysql.com/doc/refman/8.0/en/show-create-view.html) and many vendors don't even offer this functionality which I find weird (e.g. Snowflake only shows the schema for [DESCRIBE VIEW](https://docs.snowflake.net/manuals/sql-reference/sql/desc-view.html)).

I'm fine with removing `SHOW VIEW` for now?

fhueske replied:
OK, let's remove it for now and figure that out later.

fhueske commented:
Thanks for the update @twalthr. 
I left two more comments.

Thanks, Fabian

twalthr commented:
@fhueske I can only see one more comment?
